variables.tf

```diff
@@ -87,34 +87,27 @@ variable "key_protect_endpoint_type" {
   }
 }
 
-variable "existing_key_protect_instance_guid" {
+variable "existing_kms_instance_guid" {
   type        = string
-  description = "The GUID of an existing Key Protect instance, required if 'var.create_key_protect_instance' is false."
+  description = "The GUID of an existing Key Protect or Hyper Protect Crypto Services instance, required if 'var.create_key_protect_instance' is false."
   default     = null
 }
 
 variable "key_map" {
-  type        = map(list(string))
-  description = "Use this variable if you wish to create both a new key ring and new key. The map should contain the desired Key Ring name as the keys of the map, and a list of desired Key Protect Key names to create as the values for each Key Ring."
-  default     = {}
-}
-
-variable "existing_key_map" {
-  type        = map(list(string))
-  description = "Use this variable if you wish to create new keys inside already existing Key Ring(s). The map should contain the existing Key Ring name as the keys of the map, and a list of desired Key Protect Key names to create as the values for each existing Key Ring."
-  default     = {}
-}
-
-variable "force_delete_key_ring" {
-  type        = bool
-  description = "Set to `true` to force delete key ring or `false` if not"
-  default     = true
-}
-
-variable "force_delete" {
-  type        = bool
-  description = "Allow keys to be force deleted, even if key is in use"
-  default     = true
+  type = list(object({
+    key_ring_name     = string
+    existing_key_ring = optional(bool, false)
+    force_delete      = optional(bool, true)
+    keys = list(object({
+      key_name                 = string
+      standard_key             = optional(bool, false)
+      rotation_interval_month  = optional(number, 1)
+      dual_auth_delete_enabled = optional(bool, false)
+      force_delete             = optional(bool, true)
+    }))
+  }))
+  description = "A list of objects which contain the key ring name, a flag indicating if this key ring already exists, and a flag to enable force deletion of the key ring. In addition, this object contains a list of keys with all of the information on the keys to be created in that key ring."
+  default     = []
```

Comment:
suggest to add some variable validation. For example, the other internal module has this:

```
  validation {
    error_message = "Each key must have a unique name."
    condition = length(
      distinct(
        var.keys[*].name
      )
    ) == length(var.keys[*].name)
  }

  validation {
    error_message = "Rotation must be an integer between 1 and 12."
    condition = length(
      [
        for key in var.keys :
        key if key.rotation % 1 != 0 || key.rotation < 1 || key.rotation > 12
      ]
    ) == 0
  }
```

Reply:
The validation from the module level works at plan time for these variables, so they don't need to be replicated here. 

This is also one of the only downsides with the list of object style variables, doing validations at the variable level doesn't work so they have to be validated in local blocks, but in this case they're not needed because this is just a composition of modules with their own validations.

Reply:
Its better to add the validation in the variable is possible, as the error message is cleaner for consumers (what we do in locals is a hack when validation cant be done in variables). Are you saying there is a reason we cant do validation here? Can we not do what the I pasted above?

Reply:
Yes, if we try to add validations with the `[*]` expansions on objects, the checks don't actually seem to do anything, an example is doing something like 
```hcl
validation {
  condition = var.keys[*].keys[*].rotation_interval_month != 10
  error_message = "rotation interval month cannot be 10"
}
```
inside the keys variable block and then passing a rotation_interval_month value of 10 to the block doesn't fail, I'm not sure what causes this as it may be that validations don't work with splat expansions at all or it may be a quirk with using the `optional()` value on the field.

But, in this case, the module for kms-key has all the validation in place for these variables so it is not necessary to add any validation here, I tested this behavior by setting rotation_interval_month = 13 in an example and that fails at plan time.